snappy/faststreams.nim

```diff
@@ -0,0 +1,147 @@
+import
+  std/strutils,
+  pkg/faststreams/[inputs, multisync, outputs],
+  "."/[codec, encoder, exceptions],
+  ../snappy
+
+export
+  inputs, multisync, outputs, codec, exceptions
+
+{.push raises: [Defect].}
+
+proc checkCrcAndAppend(
+    output: OutputStream, data: openArray[byte], crc: uint32): bool {.
+    raises: [Defect, IOError].}=
+  if maskedCrc(data) == crc:
+    output.write(data)
+    return true
+
+proc compress*(input: InputStream, output: OutputStream) {.
+    raises: [Defect, InputTooLarge, IOError].} =
+  ## Compress all bytes of `input`, writing into `output` and flushing at the end.
+  ##
+  ## Input length must not exceed `maxUncompressedLen == 2^32-1` or
+  ## `InputTooLarge` will be raised. Other errors are raised as they happen on
+  ## the given streams.
+  doAssert input.len.isSome, "TODO: support actual .. streams"
```

Comment:
Why is this input length limitation present?
It would be easy to enforce it by maintaining a counter as you encode the blocks below.

Reply:
copied from the existing code - it's needed to compute a (temporary) output buffer that is large enough

Reply:
more accurately: if we don't have the input size, we cannot compute the maximum compressed length of the whole stream - ostensibly, with streams, we could do something like incrementally grow the output buffer (because unlike decompression, we control the block size) and this is what `getWriteableBytes` would be used for: it would give one "input block" at a time